src/tests.rs

```diff
@@ -0,0 +1,240 @@
+// modified, or distributed except according to those terms. Please review the Licences for the
+// specific language governing permissions and limitations relating to use of the SAFE Network
+// Software.
+
+#[cfg(test)]
+mod test {
+    use crate::key_gen::Phase;
+    use crate::member::NodeID;
+    use crate::Member;
+    use bincode::serialize;
+    use itertools::Itertools;
+    use quic_p2p::{Config, Peer};
+    use std::collections::{BTreeMap, BTreeSet, HashMap};
+    use std::net::{IpAddr, Ipv4Addr};
+    use std::thread::sleep;
+    use std::time::Duration;
+
+    pub const IP_LOCAL_HOST: IpAddr = IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1));
+    pub const NODE_NUM: usize = 7;
+    pub const THRESHOLD: usize = 6;
+
+    fn setup_members_and_init_dkg(non_responsives: BTreeSet<usize>) -> (Vec<NodeID>, Vec<Member>) {
+        let mut node_id_vec = vec![];
+        let mut config_vec = vec![];
+        let mut socket_addr_vec = vec![];
+        let mut peer_vec = vec![];
+        let mut map = HashMap::new();
+        let mut members = vec![];
+        let mut proposals = vec![];
+
+        // Setup Node's transport layer details
+        for _ in 0..NODE_NUM {
+            let mut config = Config::default();
+            config.ip = Some(IP_LOCAL_HOST);
+            config.port = Some(0);
+
+            let (member, socket_addr) = Member::new_for_test(config.clone()).unwrap();
+
+            let peer = Peer::Node(socket_addr);
+            let id = member.id();
+
+            node_id_vec.push(id);
+            config_vec.push(config);
+            socket_addr_vec.push(socket_addr);
+            peer_vec.push(peer);
+            members.push(member);
+        }
+
+        // Forms the group - NodeIDs and Socket Addresses.
+        for x in 0..NODE_NUM {
+            let _ = map.insert(
+                node_id_vec[x].clone(),
+                (socket_addr_vec[x], peer_vec[x].clone()),
+            );
+        }
+
+        // Initialize DKG for all the Nodes
+        for member in members.iter_mut().take(NODE_NUM) {
+            let proposal = member.init_dkg(map.clone(), THRESHOLD).unwrap();
+            proposals.push(proposal.clone());
+            std::thread::sleep(Duration::from_secs(5));
+        }
+
+        // Broadcast all the messages simultaneously via multi-threading
+        for x in 0..NODE_NUM {
+            if !non_responsives.contains(&x) {
+                // Could get delayed during connecting in real-network - keep trying in a loop
+                let mut connected = members[x].if_connected_to_all();
+                if connected {
+                    members[x].broadcast(proposals[x].clone()).unwrap();
+                } else {
+                    while !connected {
+                        connected = members[x].if_connected_to_all();
+                        if connected {
+                            members[x].broadcast(proposals[x].clone()).unwrap();
```

Comment:
It's nothing but just a check to confirm that we are connected to all the nodes before broadcasting our initial message.

Reply:
but here in addition to the check, it seems also tries to connect.
This gives a bit misleading impression that broadcasting of message is related to the connection?

Reply:
No, it doesn't try to connect, it continuously queries to see if we are connected to everybody. And once connected, it starts firing the Initial broadcast messages :)

Reply:
Well, I said `it seems tries to connect`
This function call acutally may becomes an infinite loop if never got connected.
Also, there is no `wait` (ease of thread) here during the query. which means it may occupy the whole cpu and won't give the execute cycyle to other threads.